sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,500 @@
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(getQuery() != null, "CassandraIO.read() requires a query to be set via "
+          + "withQuery(query)");
+      checkState(getEntityName() != null, "CassandraIO.read() requires an entity name to be set "
+          + "via withEntityName(entityName)");
+      checkState(getTable() != null, "CassandraIO.read() requires a table to be set via "
+          + "withTable(table)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+
+      getConnectionConfiguration().populateDisplayData(builder);
+
+      builder.addIfNotNull(DisplayData.item("query", getQuery()));
+      builder.addIfNotNull(DisplayData.item("entityName", getEntityName().getName()));
+      builder.addIfNotNull(DisplayData.item("rowKey", getRowKey()));
+      builder.addIfNotNull(DisplayData.item("table", getTable()));
+    }
+
+  }
+
+  private static class BoundedCassandraSource<T> extends BoundedSource<T> {
+
+    private Read<T> spec;
+    private String splitQuery;
+
+    BoundedCassandraSource(Read<T> spec, String splitQuery) {
+      this.spec = spec;
+      this.splitQuery = splitQuery;
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return spec.getCoder();
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions pipelineOptions) {
+      return false;
+    }
+
+    @Override
+    public BoundedReader<T> createReader(PipelineOptions pipelineOptions) {
+      return new BoundedCassandraReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+      try (Cluster cluster = spec.getConnectionConfiguration().getCluster()) {
+        try (Session session = cluster.newSession()) {
+
+          ResultSet resultSet = session.execute("SELECT partitions_count, mean_partition_size FROM "
+              + "system.size_estimates WHERE keyspace_name = ? AND table_name "
+              + "= ?", spec.getConnectionConfiguration().getKeyspace(), spec.getTable());
+
+          Row row = resultSet.one();
+
+          long partitionsCount = 0;
+          long meanPartitionSize = 0;
+          if (row != null) {
+            partitionsCount = row.getLong("partitions_count");
+            meanPartitionSize = row.getLong("mean_partition_size");
+          }
+
+          LOG.debug("Cassandra table {} (keyspace {}) estimated to {} bytes", spec.getTable(),
+              spec.getConnectionConfiguration().getKeyspace(), partitionsCount * meanPartitionSize);
+          return partitionsCount * meanPartitionSize;
+        }
+      }
+    }
+
+    @Override
+    public List<BoundedSource<T>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                PipelineOptions pipelineOptions) {
+      int exponent = 63;
```

Comment:
(just another reminder to use ByteKeyRange. Code for splitting key ranges is usually insanely tricky and I can't stress enough how much I recommend NOT to reimplement it)